Add AudioStreamPlayerInternal to unify stream players

KoBeWi opened the pull request:
We have 3 AudioStreamPlayer classes and they share a lot of code. Code duplication, especially this big, is bad and that doesn't need explaining. Every fix or common feature has to be implemented 3 times and there is risk of code differences that result in bugs.

This PR tries to unify the code under a single class - AudioStreamPlayerInternal. It's an internal class used by the 3 audio players (not by inheritance). I delegated all common code and fields to the new class. The best indicator how big was the code duplication is that this PR has more deleted lines despite I added a whole new class. I also cleaned up some includes.

I tried to avoid introducing changes, but there are small differences from the original code, mostly in the order of some lines (some things were ordered differently in different players and common methods will obviously unify that). Probably needs testing to check whether all features are still working and there is no change in behavior. I did some basic tests and the audio still plays at least.

In the meantime I also came up with an alternate approach of taking advantage of C++ multiple inheritance. Maybe I'll try that too, as it should save some more code and possibly make it more readable.

KoBeWi committed:
Add AudioStreamPlayerInternal to unify stream players

KoBeWi commented on scene/2d/audio_stream_player_2d.cpp:
```diff
@@ -33,84 +33,46 @@
+			if (!internal.stream_playbacks.is_empty() && internal.active.is_set()) {
+				internal.process();
 			}
+			internal.ensure_playback_limit();
```
The playback limit is applied each frame, unlike in AudioStreamPlayer, which does it only on `play()`. Maybe other players could do that too?

KoBeWi commented on scene/audio/audio_stream_player.cpp:
```diff
@@ -220,43 +126,19 @@ AudioStreamPlayer::MixTarget AudioStreamPlayer::get_mix_target() const {
+	internal.set_playing(p_enable);
 }
 
 bool AudioStreamPlayer::_is_active() const {
```
This method code was the same as `is_playing()`. I unified it to do `active.is_set()` as other players, idk if it makes difference.

MJacred replied:
Hm… for AudioStreamPlayer, the meaning changes from "AudioServer has my playback registered and setup, and started playing" to "I started being processed and the AudioServer is about to start setting up my playback".

But no idea if this has any tangible repercussions. `_is_active` is a Node thing afaik

MJacred replied:
Not sure if relevant, but does animation use this for syncing the start? Audio already starts earlier than other things. For reference: https://github.com/godotengine/godot/issues/82698#issuecomment-1871426859

KoBeWi commented on scene/3d/audio_stream_player_3d.cpp:
```diff
@@ -595,83 +543,52 @@ void AudioStreamPlayer3D::seek(float p_seconds) {
 void AudioStreamPlayer3D::set_bus(const StringName &p_bus) {
-	//if audio is active, must lock this
+	// If audio is active, must lock this.
 	AudioServer::get_singleton()->lock();
```
The lock does not exist in other players. Should it be removed?

ellenhp replied:
Good catch. It should be unnecessary. We handle thread contention between mixing and modification of playback params like this in AudioServer with SafeList.

Mickeon commented:
> In the meantime I also came up with an alternate approach of taking advantage of C++ multiple inheritance. Maybe I'll try that too, as it should save some more code and possibly make it more readable.

I'm very interested. Readability seems to be the must. But, is something like the suggestion ever used across  the rest of the codebase?

Whatever path is taken to unite the AudioStreamPlayers, it would be nice _in the future_ to do the same for other classes with a lot of duplicated code.

ellenhp commented:
An important discussion point is whether this may interfere with people creating custom streams with gdextension. I think the answer is no because they can just inherit audiostream and continue duplicating all the code we're consolidating here. But that's a use case I want us to continue supporting so I want to bring it up :)

Mickeon commented:
Compatibility is retained by doing this. From the outsider's perspective nothing should change.

KoBeWi commented:
> But, is something like the suggestion ever used across the rest of the codebase?

No, AFAIK every class is single inheritance.

I tried that approach and seems like you can't bind methods from another class (in that case AudioStreamPlayerInternal would be the secondary parent class), which makes it useless. Maybe there is a way to untangle the inheritance, but it's not that easy, so meh. The solution in this PR is better than what we have anyway.

ellenhp commented:
> Compatibility is retained by doing this. From the outsider's perspective nothing should change.

Yeah makes sense, I'm just trying to be thorough. I also think we should come up with a better name because we have a couple audio streams that won't end up subclassing this new internal class because they're dynamic (microphone, generator, probably others).

Mickeon commented:
It could be called something like "Core". I had it for a bit and it's the only thing that comes to mind.

KoBeWi commented:
> I also think we should come up with a better name because we have a couple audio streams that won't end up subclassing this new internal class

The new class is for audio stream players (it's AudioStream**Player**Internal), not audio streams. I think the name is fine.

ellenhp commented:
Yeah sorry, neither of my comments make any sense for this. I got a covid vaccine about 36 hours ago and I'm still in the living dead phase of recovery. Didn't help that I hadn't looked at the code yet. I do want to refactor streams at some point too but there's a lot less we can do there without breaking stuff. As you noted, you can't create a new player in GDExtension yet so we really don't have to worry about breaking compat, and I'm not even aware of any custom modules that implement their own players without going through an entirely different middleware.

MJacred commented on scene/audio/audio_stream_player_internal.cpp:
```diff
@@ -0,0 +1,253 @@
+	physical = p_physical;
+
+	AudioServer::get_singleton()->connect("bus_layout_changed", callable_mp((Object *)node, &Object::notify_property_list_changed));
+	AudioServer::get_singleton()->connect("bus_renamed", callable_mp((Object *)this, &Object::notify_property_list_changed).unbind(3));
```
I'm not sure about this, but shouldn't the second connect also be to `node`?

```suggestion
	AudioServer::get_singleton()->connect("bus_renamed", callable_mp((Object *)node, &Object::notify_property_list_changed).unbind(3));
```

KoBeWi replied:
Yes, I already fixed it locally.

MJacred commented on scene/2d/audio_stream_player_2d.h:
```diff
@@ -64,12 +63,6 @@ class AudioStreamPlayer2D : public Node2D {
-
 	void _set_playing(bool p_enable);
 	bool _is_active() const;
 
```
You forgot to remove a few lines below:

```c++
	void _on_bus_layout_changed();
	void _on_bus_renamed(int p_bus_index, const StringName &p_old_name, const StringName &p_new_name);
```

MJacred commented on scene/3d/audio_stream_player_3d.cpp:
```diff
@@ -595,83 +543,49 @@ void AudioStreamPlayer3D::seek(float p_seconds) {
-	AudioServer::get_singleton()->lock();
-	bus = p_bus;
-	AudioServer::get_singleton()->unlock();
+	internal.bus = p_bus;
```
```suggestion
	internal.bus = p_bus; // This will be pushed to the audio server during the next physics timestep, which is fast enough.
```

And could you add this info to the AudioStreamPlayer2D/3D docs, please?

Mickeon replied:
Somewhat out of scope of the PR since it's always been the case, pretty much

MJacred replied:
The comment was copied from same func from AudioStreamPlayer2D.

The docs update is just for info for users to not be confused about possible delays if their physics process is stuttering

KoBeWi replied:
I'd wait until a confused user opens an issue. So far it hasn't been a problem to anyone, likely because changing buses dynamically is a very uncommon action.
I copied the comment though.

MJacred reviewed (changes requested):
The rest should be fine, I'd say. Some processing order was changed, e.g. in `_notification`, but should make no difference (I think).

KoBeWi commented on scene/3d/audio_stream_player_3d.cpp:
```diff
@@ -242,32 +243,12 @@ void AudioStreamPlayer3D::_notification(int p_what) {
 			AudioServer::get_singleton()->remove_listener_changed_callback(_listener_changed_cb, this);
 		} break;
 
-		case NOTIFICATION_PREDELETE: {
-			stop();
-		} break;
```
This is handled by the internal class, which also receives this notification.

MJacred replied:
How?
The relevant internal class (see code below) cannot reset `setplay`. That variable only exists in AudioStreamPlayer2D and *3D.

```
void AudioStreamPlayerInternal::notification(int p_what) {
...
            case Node::NOTIFICATION_PREDELETE: {
                       for (Ref<AudioStreamPlayback> &playback : stream_playbacks) {
                            AudioServer::get_singleton()->stop_playback_stream(playback);
                        }
                        stream_playbacks.clear();
            } break;
...
}
```

KoBeWi replied:
Ok fixed (not pushed yet).

KoBeWi commented on scene/2d/audio_stream_player_2d.cpp:
```diff
@@ -33,84 +33,47 @@
 			AudioServer::get_singleton()->remove_listener_changed_callback(_listener_changed_cb, this);
 		} break;
 
-		case NOTIFICATION_PREDELETE: {
-			stop();
-		} break;
```
See below.

KoBeWi commented on scene/audio/audio_stream_player_internal.cpp:
```diff
@@ -0,0 +1,253 @@
+			for (Ref<AudioStreamPlayback> &playback : stream_playbacks) {
+				AudioServer::get_singleton()->stop_playback_stream(playback);
+			}
+			stream_playbacks.clear();
```
This is code from original AudioStreamPlayer, which did only this. I'd call `stop()` if I wanted to do more, but unified it into this part. The flag is irrelevant at that point.

tbh this made me realize that `setsetplay` is irrelevant too.

MJacred replied:
> This is code from original AudioStreamPlayer, which did only this

True. But AudioStreamPlayer2D & *3D cleared `active`, because they called `stop()`.

Whether the flag is relevant or not. No idea. At least, afaik, the audio system does not use that.

Similar with `setplay`.

I don't know how `NOTIFICATION_PREDELETE`, `AudioStreamPlayerInternal::_set_process` and `AudioStreamPlayerInternal::process` exactly interact with each other. The order of execution. And what happens if the 2d/3d audio player node is called with `play()` and then immediately deleted. Or similar edge cases.

The feedback regarding `active` and `setplay` is about preserving the original behaviour - and make it consistent among the audio players.

No idea if they were a "just in case" measure or if they actively prevented edge cases. So, if you know they don't matter, then it's completely fine with me to not reset these variables.

KoBeWi replied:
`NOTIFICATION_PREDELETE` is received just before the object is freed. The object does not call any method anymore (except destructor), even deferred calls.

MJacred replied:
Ah, then as long as `NOTIFICATION_INTERNAL_PHYSICS_PROCESS` is not executed afterwards, then it sounds safe. Thanks!

KoBeWi commented:
Rebased and updated. I had to change the AudioStreamPlayerInternal to Object, to handle the newly added `parameter_list_changed` signal. I also fixed the `null` issue from #86473.

reduz reviewed (approved):
Looks good, I was hopeful it would save more code, but I guess its still a lot less error prone and easier to maintain. Great work!

akien-mga commented:
Failing test:
```
./tests/core/object/test_class_db.h:833:
TEST SUITE: [ClassDB]
TEST CASE:  [ClassDB] Add exposed classes, builtin types, and global enums
  [ClassDB] Validate exposed classes

./tests/core/object/test_class_db.h:376: ERROR: CHECK_FALSE( (p_arg.name.is_empty() || p_arg.name.begins_with("_unnamed_arg")) ) is NOT correct!
  values: CHECK_FALSE( true )
  logged: Unnamed argument in position 0 of method 'AudioStreamPlayer2D.set_pitch_scale'.
```

YuriSizov commented on scene/audio/audio_stream_player_internal.h:
```diff
@@ -0,0 +1,105 @@
+class AudioStreamPlayback;
+class Node;
+
+class AudioStreamPlayerInternal : public Object {
+	GDCLASS(AudioStreamPlayerInternal, Object);
```
Do we actually need this to be a Godot Object? Can it be just a class?

KoBeWi replied:
It needs to be Object to connect the signal.

YuriSizov replied:
Ah, now I see it! Okay then.

YuriSizov reviewed (approved):
Gave it a quick look, and it looks good!

akien-mga commented:
Thanks!
